design-decisions/2018-11-14-Heating-Cooling-Representation.md

```diff
@@ -0,0 +1,125 @@
+3+4 ~> 3+[]      -- heating
+7 ~> 3+[]        -- rewriting
+3+7              -- cooling
+10               -- rewriting
+```
+
+In many cases it is desirable to allow expressions to be evaluated in a fairly
+free order. Consider the following C example:
+```
+(f() + g()) + h()
+```
+where `f`, `g` and `h` have side effects. According to the standard, these
+functions can be called in any order, e.g. it is valid to call `g`, then `h`,
+then `f`.
+
+In all-path reachability proofs one attempts to show that, when starting from
+a given pattern, no matter how it is evaluated, one reaches a certain
+target pattern.
+
+In coinductive proofs, whenever there is a cycle in which a pattern can be
+rewritten, indirectly or not, to itself, anything can be proven.
+
+Problem/Questions
+-----------------
+
+How to encode the heating and cooling rules in kore, in such a way that there
+are no rewrite cycles, we can compute all patterns reachable from a given one,
+and we can prove that only those are reachable, while allowing any evaluation
+order for the heated/cooled terms when that's desirable.
+
+Decision: Use rewriting rules with "changed" bit
+------------------------------------------------
+
+Use rewriting rules like the following, where `(,)` is a constructor,
+and its second parameter is a "changed" bit:
+```
+(a+b,x) => (a,0) ~> ([]+b,x)     if a is not a value
+(a, 1) ~> ([]+b, x) => (a+b, 1)
+(a+b, x) => (a +Int b, 1)        if a and b are values
+```
+
+Reasoning and other options considered
+--------------------------------------
+
+Options that were considered:
+
+### Using plain rewriting rules
+
+Using rules like the following generates rewrite cycles:
+```
+a+b => a ~> []+b
+a+b => b ~> a+[]
+a ~> []+b => a+b
+b ~> a+[] => a+b
+```
+
+### Using rewriting rules with side conditions
+
+```
+a+b => a ~> []+b   if a is not a value
+a+b => b ~> a+[]   if b is not a value
+a ~> []+b => a+b   if a is a value
+b ~> a+[] => a+b   if b is a value
+```
+
+Side conditions solve the rewrite cycle issue, but do not allow all possible
+expression evaluations, e.g. it does not allow evaluating `g`, then `h`, then
+`f` in the Background section example.
+
+### Using equations, e.g.
+```
+a+b = a ~> []+b
+a+b = b ~> a+[]
+```
+
+These solve the evaluation order issue in an elegant way, and it is
+computationally feasible to compute all possible expression expansions.
+
+However, it seems rather difficult to prove that one needs to consider only
+rewrites starting from those expansions if one wants to find if we reach the
+target pattern no matter how we evaluate and rewrite the start pattern.
+
+While we could, maybe, ignore this issue, it seems kind of bad.
+
+### Using rewriting rules with "changed" bit.
+
+```
+(a+b,x) => (a,0) ~> ([]+b,x)     if a is not a value
+(a+b,x) => (b,0) ~> (a+[],x)     if a is not a value
+(a, 1) ~> ([]+b, x) => (a+b, 1)
+(b, 1) ~> (a+[], x) => (a+b, 1)
+(a+b, x) => (a +Int b, 1)        if a and b are values
```

Comment:
About proving: We might define this as the meaning of heating and cooling and then we wouldn't need to prove anything. Even if we need to prove it, it will probably be easier than providing automatic proofs for all-path reachability with equations.

Reply:
I switched the proposed solution to using equations. How does it look now?